Automattic/jetpack
# Refactor/add products v1 to v2 proxied endpoint

Fixes 106-gh-automattic/gold
Blocks https://github.com/Automattic/wp-calypso/pull/80205
Blocks D118667-code

## Proposed changes:

* Adds update and delete API endpoints for V2
* Proxies these requests to WPCOM from the Jetpack site
* Handles these requests if IS_WPCOM.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






This change involves WPCOM and requires access to a WPCOM sandbox to test.

There are three components to this issue:
- This patch for Jetpack which adds the "update" and "delete" endpoints for memberships plans to the backend. This patch could be tested by itself if it is downloaded to sandbox, public-api is directed to sandbox and the endpoints are called with the api developer console. 
- Calypso patch https://github.com/Automattic/wp-calypso/pull/80205 which calls the newly added "update" and "delete" endpoints from calypso. This patch could be tested by itself to see that endpoints were hit, but requires the Jetpack patch to function correctly - otherwise those endpoints would result in 404s, because they are defined in the Jetpack.
- WPCOM D118667-code which removes (now unused) the v1.1 endpoints from WPCOM. This patch could be tested by itself by patching sandbox, directing public-api to sandbox and using the api developer console to confirm the v1.1 endpoints for membership products no longer exist. Instead, this test plan uses this to help assert the v1.1 endpoints aren't being called because they no longer exist.

### Testing Non-Proxied Flow

Currently, the only flow available on production.

#### Testing from Calypso Localhost / Jetpack Source
1. Patch D118667-code and sandbox public API.
2. Checkout this branch locally `git fetch && git checkout origin refactor/add-products-v1-to-v2-proxied-endpoint`
3. Build from source with `pnpm jetpack build plugins/jetpack`
4. Sync to your D118667-patched sandbox with `pnpm jetpack rsync jetpack`
5. Make sure the public api is pointing to your sandbox
6. Clone the `wp-calypso` repo if you haven't already, and checkout `refactor/memberships-api-calls-to-v2`
7. Build calypso with `yarn` and then `yarn start`.
8. Open http://calypso.localhost:3000 in a browser and wait for calypso to finish building.
9. Once calypso loads, append `/earn/payment-plans` to the end of the url and hit enter. If you get "You are not authorized to view this page" click "switch site" and select your Jetpack site.
8. You should now be at the calypso.localhost version of `/earn/payment-plans/<your jetpack domain>`. If this is a new Jetpack site, you'll be greeted with an empty list of payment plans.
9. Click to add a new payment plan. The plan should add successfully.
10. Click to edit the created payment plan, make some adjustments and click save. The plan should save successfully and your edits should persist if you refresh the page.
11. Click to delete the created payment plan. The plan should delete, and if you refresh the page the plan should not re-appear in the list of payment plans.
12. Return to your Jetpack site and create a new post, this time with a payment button. Click the icon to select the payment plan and you should see a list of payment plans. Click add new payment plan and fill out the form, clicking save when done. The plan should exist when you return to the calypso.localhost page and view your payment plans.

#### Testing with Calypso Live / Jurassic Ninja

It's possible to test with Jurassic Ninja, but please note that currently no AJAX/XHR requests are expected to go to through to the Jurassic Ninja site's backend - instead all AJAX/XHR requests go to public-api directly from the frontend, meaning that the WPCOM sandbox will be processing the backend requests. Mostly, this prevents needing to build Jetpack and Calypso by source.

1. Patch D118667-code and sandbox public API.
2. Use the release branch in Jetpack Beta on your Jetpack site. In the upper-right corner of the wp-admin interface of your Jetpack site, click Jetpack Beta, then the first option, then "Manage". Paste the following into the Release branch option and click "Activate" when the result is found for: refactor/add-products-v1-to-v2-proxied-endpoint.
3. Download the patched version of Jetpack to your sandbox with `bin/jetpack-downloader test jetpack refactor/add-products-v1-to-v2-proxied-endpoint`.
4. On your Jetpack site, connect Jetpack to your WPCOM account, and add a new post with the "Subscribe" payment block.
5. Follow the prompts to setup a payment plan. Making note of if and when you are transferred to wordpress.com. If you see this, do not continue on wordpress.com - instead, continue to the next step, where we will use the patched Calypso to create a new payment plan.
6. Open calypso via the Calypso live links on https://github.com/Automattic/wp-calypso/pull/80205
7. Once calypso loads, append `/earn/payment-plans` to the end of the url and hit enter. If you get "You are not authorized to view this page" click "switch site" and select your Jetpack site.
8. You should now be at the calypso.live version of `/earn/payment-plans/<your jetpack domain>`. If this is a new Jetpack site, you'll be greeted with an empty list of payment plans.
9. Click to add a new payment plan. The plan should add successfully.
10. Click to edit the created payment plan, make some adjustments and click save. The plan should save successfully and your edits should persist if you refresh the page.
11. Click to delete the created payment plan. The plan should delete, and if you refresh the page the plan should not re-appear in the list of payment plans.
12. Return to your Jetpack site and create a new post, this time with a payment button. Click the icon to select the payment plan and you should see a list of payment plans. Click add new payment plan and fill out the form, clicking save when done. The plan should exist when you return to the calypso live page and view your payment plans.
